### DISCUSSION TITLE
Issues with Bonsai to Arduino Trigger

### DISCUSSION THREAD

**Discussion by nazdundar** [bonsai-rx/bonsai#1633] [Q&A]
Hi all,

I'm trying to use Bonsai to implement Real-Time Place Preference in our lab and having issues with establishing a connection between Bonsai and Arduino. I've tried reading all relevant posts. I am able to detect the mouse's presence in a specific ROI but there seems to be a disconnect between Bonsai and Arduino such that the laser does not come on with roi detection. I installed StandardFirmata to my Arduino Uno Board and I'm attaching all the files I've been using. Could anybody help me figure this out? Thanks!
![Screenshot 2023-11-20 135650](https://github.com/bonsai-rx/bonsai/assets/40470122/08a6e5a0-c3f7-4584-9129-1267b09f1867)

Arduino Code:
```
int input = 8;
int output = 12;
int inputState = 0;

// the setup function runs once when you press reset or power the board
void setup() {
  // initialize digital pin 13 as an output.
  pinMode(input, INPUT);
  digitalWrite(input,LOW);
  pinMode(output, OUTPUT);
  digitalWrite(output,LOW);
  Serial.begin(57600);
}

// the loop function runs over and over again forever
void loop() {
  inputState = digitalRead(input);
  Serial.println (inputState);
  if (inputState == HIGH) {
      digitalWrite(output, HIGH);   // turn the LED on (HIGH is the voltage level)
      delay(2*1000);              // wait for a second
      digitalWrite(output, LOW);    // turn the LED off by making the voltage LOW
      delay(3*1000);              // wait for a second
  }
  else {
    digitalWrite(output, LOW);
  }
}
```
``

**Comment by bruno-f-cruz** [comment]
Heya!
I am not sure I understand what do you mean by installed firmata. Firmata is a sketch like the one you posted, so if you are uploading your sketch, firmata will be overridden by your code and bonsai won't be able to interface with it.

**Reply by nazdundar** [reply]
Hi,
Thank you so much for answering! I'm really new to both bonsai and arduino so I assumed uploading the StandardFirmata sketch and my sketch after it would have been fine! Is there another approach I should use? Can I just copy paste this code to the end of the StandardFirmata sketch? Thanks again for all your help!

**Reply by bruno-f-cruz** [reply]
If you just want to trigger a digital output as your bonsai workflow suggests, you don't need to upload your sketch at all. Firmata will do the trick.

**Reply by nazdundar** [reply]
How would I ensure that the arduino triggers my laser in a specific manner? If I wanted to do 20Hz stimulation 2 sec on 3 sec off, how would I get firmata to have that?

**Reply by bruno-f-cruz** [reply]
For temporal high-precision applications you cant rely on Bonsai to generate your signal. So, you have at least 2 options:

1 - Change the firmata software to trigger your protocol of choice when triggering a specific pin/message.
2- The easiest would be to probably code your own state machine in Arduino that is waiting for a specific serial message from the PC host that you can send using the Serial package in Bonsai. This last option would not use firmata nor the arduino package, but it would require you to code your own arduino sketch to handle the serial communication and laser protocol.

**Comment by HildeGunnink** [comment]
Hi Bruno, I just found this thread today. One of your suggenstions is actually what I am currently trying to do: I have written my own Arduino code that should be triggered by a serial command as input. However, I am quite new to Bonsai, and I don't udnerstand how to transform the measured values from the CameraCapture pipeline to a 1 or 0 output towards Arduino. My current setup in Bonsai is:
CameraCapture - Crop - Sum - Val0 - LessThan - 
and from here I need to do a conversion to 1 and 0 strings, and then finally a serialstringwrite to arduino on COM3 at 57600 baudrate. However, I just cant seem to find the operators that I need in Bonsai, which I think (based on ChatGPT) are: ConditionalOperator and SerialStringWrite. Do you have experience with this? Thank you so much in advance!

**Reply by bruno-f-cruz** [reply]
Does this answer your question?

https://bonsai-rx.org/docs/tutorials/acquisition.html#exercise-7-control-an-led

**Reply by HildeGunnink** [reply]
Hi Bruno, thank you for your reply!

I actually tried that first, but it didn't work. To my understanding of the documentation, DigitalOutput writes directly to an Arduino output pin, meaning that the script sending the right modulated output could be bypassed. In any case, it didn't work :) Currently trying via SerialWriteLine with a floating CreateSerialPort, but also that does not seem to trigger the Arduino. I have tested the Arduino script through the Serial Monitor tool in the Arduino IDE and it can read and write the right commands, and the script works as it should in response to the commands. The issue really seems to be how to get Bonsai to send the signal to Arduino in a way that Arduino reads it, and for some reason the Serial package tools I am using, don't seem to work

**Reply by HildeGunnink** [reply]
Tried changing Dtr to True for the CreateSerialPort node based on: https://groups.google.com/g/bonsai-users/c/IbLk_JnqEwc, that seemed to have worked; now at least there is some communication from Bonsai being read to Arduino. The information Arduino received was garbled, despite being the Baudrate for the Arduino Serial.read and Bonsai CreateSerialPort being identical.

**Reply by bruno-f-cruz** [reply]
The DTR pin will be needed depending on the Arduino board you are using. The old UNO board did not require it to be set, but the DUE, for instance, does. Not sure which board you have so it is likely it may need to be set in your case.

Regarding converting a bool to a string, the easiest way is to use an `ExpressionTransform` with a ternary operator `it ? "On" : "Off"`. Alternatively you can also use the conditional logic exemplified here (https://bonsai-rx.org/docs/tutorials/state-machines.html#exercise-5-specifying-conditional-task-outcomes) and use the `String` operator instead of the `SelectMany` used in the example.

Finally, I will also note that the `Format` operator may also be useful in some cases, but I will that one for you to try.

Hope it helps!

**Reply by HildeGunnink** [reply]
I am using an Arduino Uno, but nevertheless it solved the problem! 

For future reference, below is our Bonsai setup: 
CameraCapture - Sample - Crop - Grayscale - Threshold - Sum - Val0 - LessThan - SerialWriteLine (NewLine set to \n and PortName should be the COM port for your arduino and computer (can be found in the Arduino IDE when Arduino is connected)), and a floating CreateSerialPort (PortName should match Arduino and SerialWriteLine port, Baudrate should match Arduino script, DtrEnable set to True, NewLine set to \n).

note, this pipeline outputs 'True' and 'False' to Arduino, so Arduino read commands have to be read as True and False;

For the Arduino script the main things for reading the commands are then: 
```
void setup() {
  Serial.begin(57600); //this is the baudrate, so whatever matches the one in Bonsai
}

void loop() {
  if (Serial.available() > 0) {
    String command = Serial.readStringUntil('\n'); // note this matches with Bonsai's NewLine
    command.trim();
    
    if (command == "True" || command == "true") {
      }  

    else if (command == "False" || command == "false") {
    }
}
}
```
